File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/text/ComponentTranslate.java
package pize.tests.voxelgame.main.text;

import pize.graphics.util.color.Color;
import pize.tests.voxelgame.main.lang.Language;

public class ComponentTranslate extends Component{
    
    private final String translationKey;
    private final Component[] arguments;
    private Language translatedWith;
    
    public ComponentTranslate(TextStyle style, Color color, String translationKey, Component... arguments){
        super(style, color);
        this.translationKey = translationKey;
        this.arguments = arguments;
    }
    
    
    public void updateTranslation(Language language){
        if(language == translatedWith)
            return;
        
        final String translation = language.getTranslation(translationKey);
        final String[] translationParts = translation.split("%s");
        
        int i = 0;
        super.formattedText(translationParts[0]);
        for(Component argument: arguments){
            super.addComponent(argument);
            super.formattedText(translationParts[++i]);
        }
        
        translatedWith = language;
    }
    
    
    @Override
    public String toString(){
        return super.toString();
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/text/ComponentTranslate.kt
package pize.tests.voxelgame.main.text

import pize.graphics.util.color.Color
import pize.tests.voxelgame.main.lang.Language

class ComponentTranslate(
    style: TextStyle?,
    color: Color,
    private val translationKey: String?,
    vararg arguments: Component
) : Component(style, color) {
    private val arguments: Array<Component>
    private var translatedWith: Language? = null

    init {
        this.arguments = arguments
    }

    fun updateTranslation(language: Language) {
        if (language === translatedWith) return
        val translation = language.getTranslation(translationKey)
        val translationParts = translation!!.split("%s".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var i = 0
        super.formattedText(translationParts[0])
        for (argument in arguments) {
            super.addComponent(argument)
            super.formattedText(translationParts[++i])
        }
        translatedWith = language
    }

    override fun toString(): String {
        return super.toString()
    }
}
